openssl/src/pkey_ml_dsa.rs

```diff
@@ -0,0 +1,254 @@
+//! Module-Lattice-Based Digital Signatures.
+//!
+//! ML-DSA is a signature algorithm that is believed to be secure
+//! against adversaries with quantum computers. It has been
+//! standardized by NIST as [FIPS 204].
+//!
+//! [FIPS 204]: https://csrc.nist.gov/pubs/fips/204/final
+
+use crate::error::ErrorStack;
+use crate::ossl_param::{OsslParamArray, OsslParamBuilder};
+use crate::pkey::{PKey, Private, Public};
+use crate::pkey_ctx::PkeyCtx;
+use crate::{cvt, cvt_p};
+use foreign_types::ForeignType;
+use libc::c_int;
+use openssl_macros::corresponds;
+use std::ffi::CStr;
+use std::marker::PhantomData;
+use std::ptr;
+
+// Safety: these all have null terminators.
+// We cen remove these CStr::from_bytes_with_nul_unchecked calls
+// when we upgrade to Rust 1.77+ with literal c"" syntax.
+
+const OSSL_PKEY_PARAM_SEED: &CStr = unsafe { CStr::from_bytes_with_nul_unchecked(b"seed\0") };
+const OSSL_PKEY_PARAM_PUB_KEY: &CStr = unsafe { CStr::from_bytes_with_nul_unchecked(b"pub\0") };
+const OSSL_PKEY_PARAM_PRIV_KEY: &CStr = unsafe { CStr::from_bytes_with_nul_unchecked(b"priv\0") };
+const MLDSA44_CSTR: &CStr = unsafe { CStr::from_bytes_with_nul_unchecked(b"ML-DSA-44\0") };
+const MLDSA65_CSTR: &CStr = unsafe { CStr::from_bytes_with_nul_unchecked(b"ML-DSA-65\0") };
+const MLDSA87_CSTR: &CStr = unsafe { CStr::from_bytes_with_nul_unchecked(b"ML-DSA-87\0") };
+
+const MLDSA44_STR: &str = "ML-DSA-44";
+const MLDSA65_STR: &str = "ML-DSA-65";
+const MLDSA87_STR: &str = "ML-DSA-87";
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum Variant {
+    MlDsa44,
+    MlDsa65,
+    MlDsa87,
+}
+
+impl Variant {
+    pub(crate) fn as_str(&self) -> &'static str {
+        match self {
+            Variant::MlDsa44 => MLDSA44_STR,
+            Variant::MlDsa65 => MLDSA65_STR,
+            Variant::MlDsa87 => MLDSA87_STR,
+        }
+    }
+
+    pub(crate) fn as_cstr(&self) -> &'static CStr {
+        match self {
+            Variant::MlDsa44 => MLDSA44_CSTR,
+            Variant::MlDsa65 => MLDSA65_CSTR,
+            Variant::MlDsa87 => MLDSA87_CSTR,
+        }
+    }
+}
+
+/// Generate an ML-DSA PKey.
+#[corresponds(EVP_PKEY_fromdata)]
```

Comment:
I don't think this corresponds is particularly useful, it's too generic. We can drop it. (Same for all others like this)